### PR TITLE
[RSDK-3695] Implement a control Loop for SetVelocity in sensor-controlled base

### PR DESCRIPTION
**Pull request by randhid**
Keeping it in draft because I got to a place where I was stuck.

Questions:
- [x] Is the simple control loop configured correctly? Will the endpoint handle both setting and getting of state?
- [x] ~I think we need to start the loop every SetVelocity call as well as in the creator. We need to stop it in every other API call, then.~ Nope.
- [x] ~This follows a dual control loop configuration on linear and angular velocity - I am a bit unclear as to how the set-point will get updated when a SetVelocity command is issued.~
- [x] Investigate and fix panic condition in control package creator when server is running.

Tasks:
- [x] Write hardware based test harness using a robot configured with this base. Evaluate linear and angular speeds against motor ticks and imu readings for a few trajectories. [RDSK-5356, external repo]
- [x] Generate trajectories for the above task. [RDSK-5356, external repo]
- [x] Write software tests for SetVelocity using this loop, to check for proper behavior.
- [x] Follow-up ticket to migrate Spin with all the angular checks into its own controlled end-point.

To reviewers: no rush! I want to take the time and merge this in properly, if it needs to wait till I'm back for hardware tests that's okay. Otherwise I'll coordinate with you for testing while I'm away.

### CODE REVIEW

File: components/base/sensorcontrolled/sensorcontrolled.go [hunk 1]
```diff
@@ -176,33 +217,13 @@ func (sb *sensorBase) MoveStraight(
 	ctx, done := sb.opMgr.New(ctx)
 	defer done()
 	sb.setPolling(false)
-	return sb.controlledBase.MoveStraight(ctx, distanceMm, mmPerSec, extra)
```
Comment (randhid) [inline components/base/sensorcontrolled/sensorcontrolled.go:179]: Would the reviewers like me to move this in a separate pr for clarity?

File: components/base/sensorcontrolled/sensorcontrolled.go [hunk 2]
```diff
@@ -251,12 +311,24 @@ func (sb *sensorBase) SetPower(
 ) error {
 	sb.opMgr.CancelRunning(ctx)
 	sb.setPolling(false)
+	if sb.angularVelocityLoop != nil {
+		sb.angularVelocityLoop.Stop()
```
Comment (randhid) [inline components/base/sensorcontrolled/sensorcontrolled.go:315]: [q] Does this make the loop nil? What is the difference between non-nil and started?

File: components/base/sensorcontrolled/velocities.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	"go.viam.com/rdk/utils"
+)
+
+type linearVelControllable struct {
+	control.Controllable // is this necessary?
+	sb                   base.Base
+	ms                   movementsensor.MovementSensor
+}
+
+func (l *linearVelControllable) SetState(ctx context.Context, state float64) error {
```
Comment (randhid) [inline components/base/sensorcontrolled/velocities.go:23]: [q] These are two separate loops, I may need to combine the set state function or have a helper function feed into the unexported function called by SetVelocity, so not to zero-out a linear call when an angular is commanded.

File: components/base/sensorcontrolled/sensorcontrolled.go [hunk 4]
```diff
@@ -195,10 +241,24 @@ func (sb *sensorBase) SetVelocity(
 	var sensorCtx context.Context
 	sensorCtx, sb.sensorLoopDone = context.WithTimeout(context.Background(), timeOut)
 
+	if err := sb.angularVelocityLoop.Start(); err != nil {
```
Comment (npmenard) [inline components/base/sensorcontrolled/sensorcontrolled.go:244]: why so many  starts?

File: components/base/sensorcontrolled/velocities.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	return vel.Y, nil
+}
+
+type angularVelControllable struct {
+	control.Controllable
+	sb base.Base
+	ms movementsensor.MovementSensor
+}
+
+func (l *angularVelControllable) SetState(ctx context.Context, state float64) error {
```
[the remark this answers is not in the document]
Comment (npmenard) [reply components/base/sensorcontrolled/velocities.go:41]: why not using a single loop ?

File: components/base/sensorcontrolled/sensorcontrolled.go [hunk 6]
```diff
@@ -151,6 +159,38 @@ func (sb *sensorBase) Reconfigure(ctx context.Context, deps resource.Dependencie
 		return errors.Wrapf(err, "no base named (%s)", newConf.Base)
 	}
 
+	if sb.velocities != nil {
+		sb.lvctrl = &linearVelControllable{
```
Comment (npmenard) [inline components/base/sensorcontrolled/sensorcontrolled.go:163]: i think you should have one loop if both controls the same thing

File: components/base/sensorcontrolled/sensorcontrolled.go (hunk 2 above)
Comment (npmenard) [reply components/base/sensorcontrolled/sensorcontrolled.go:315]: no, but unclear if you can restart it after stopping

File: components/base/sensorcontrolled/velocities.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (randhid) [reply components/base/sensorcontrolled/velocities.go:41]: So here's the architecture I'm trying to implement, since the two endpoints are different - we have an angular velocity sensor and actuator in one endpoint and an linear velocity sensor and actuator in another - both have desired states set by the API and command signals sent by the loop. 
IMG_2445

File: components/base/sensorcontrolled/sensorcontrolled.go (hunk 6 above)
Comment (randhid) [reply components/base/sensorcontrolled/sensorcontrolled.go:163]: They don't control the same thing - one is angular, one is linear.

File: components/base/sensorcontrolled/sensorcontrolled.go (hunk 4 above)
Comment (randhid) [reply components/base/sensorcontrolled/sensorcontrolled.go:244]: If I have a loop running in the background in this driver - my reading of the control loop code is that it will interfere with other API commands that are called independently. I do not want the control loop to be "active" and interfering with `MoveStraight`, `Spin`, or `SetPower` if linear and angular velocity sensors exist and are being polled.

I can set a flag and return immediately from the `SetState` function I suppose - like so:
```go
type endpointBase struct {
          ....
          active bool
          ....
}

func (sb *sensorBase) setPolling(...){
         ....
         sb.endpointBase.active = false
         .....
}


func (endpointBase) SetState(...) error {
         if !endpointBase.active {
         // return immediately do not set the state of the base.
         return nil
         }
         
         .....
         endpointBase.SetVelocity(...)
}

```

Or something that follows this code.

Comment (randhid) [reply components/base/sensorcontrolled/sensorcontrolled.go:244]: The issue is - there's already a polling loop in spin - one that can be converted to its loop as well - the endpoint for that would be a SetVelocity command and a State that polls an orientation sensor, with some math applied to the orientation signal to figure out if we're at target or overshot after crossing the first-fourth quadrant boundary. That logic is not in the `sensorbase/spin.go` file.

File: components/base/sensorcontrolled/sensorcontrolled.go (hunk 6 above)
Comment (gvaradarajan) [reply components/base/sensorcontrolled/sensorcontrolled.go:163]: ~~But the loops have exactly the same inputs and (including the control blocks). Am I missing something?~~

I see that the base wrappers are different, never mind

File: components/base/sensorcontrolled/sensorcontrolled.go [hunk 7]
```diff
@@ -71,6 +72,11 @@ type sensorBase struct {
 	allSensors  []movementsensor.MovementSensor
 	orientation movementsensor.MovementSensor
 	velocities  movementsensor.MovementSensor
+
+	lvctrl              *linearVelControllable
```
Comment (gvaradarajan) [inline components/base/sensorcontrolled/sensorcontrolled.go:76]: I think writing out the words instead of "lv" and "av" would be better (or at least "linVel" and "angVel"). Later on in the file the variables don't feel clear at a glance

File: components/base/sensorcontrolled/velocities.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (gvaradarajan) [reply components/base/sensorcontrolled/velocities.go:41]: Should MoveStraight and Spin just be considered special cases of SetVelocity (as in they should call SetVelocity with special inputs)? Then you could maybe get away with one control loop but more complicated math? But I suppose this might mean what's in the green square now needs to be handled by the control loop and not the dependent base code?

[the remark this answers is not in the document]
Comment (gvaradarajan) [reply components/base/sensorcontrolled/velocities.go:41]: Oh I just realized that only a single state parameter is passable (even though you mentioned it above). My bad

[the remark this answers is not in the document]
Comment (gvaradarajan) [reply components/base/sensorcontrolled/velocities.go:41]: I have an idea for how to coordinate the linear and angular velocity states, but I'm having trouble communicating it here. Maybe we could talk offline?

**Review by gvaradarajan — DISMISSED** [review]
I reviewed this as best I could, but I think there are some more questions I need to ask you offline, if that's ok?

File: components/base/sensorcontrolled/sensorcontrolled.go (hunk 7 above)
Comment (randhid) [reply components/base/sensorcontrolled/sensorcontrolled.go:76]: since this is a draft pull request I'm resolving the comments on names and styling. Will revisit once we're at that stage.

File: components/base/sensorcontrolled/velocities.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (randhid) [reply components/base/sensorcontrolled/velocities.go:41]: > Should MoveStraight and Spin just be considered special cases of SetVelocity (as in they should call SetVelocity with special inputs)? Then you could maybe get away with one control loop but more complicated math? But I suppose this might mean what's in the green square now needs to be handled by the control loop and not the dependent base code?

Yes, not in this pr.

[the remark this answers is not in the document]
Comment (randhid) [reply components/base/sensorcontrolled/velocities.go:41]: > I have an idea for how to coordinate the linear and angular velocity states, but I'm having trouble communicating it here. Maybe we could talk offline?

Yeah, sometime next week.

**Comment by randhid** [comment]
I think this is a good stopping point for this pr:
- No panics or hangs in the server
- The control loop use is under a build flag until full testing can be done. This allows the previous sensor-controlled base model functionality while this work is ongoing.
- A way to set multiple signals and get multiple signals. More control package work needs to be done, but I will try to unblock @martha-johnston's work by adding the changes now - the tests all pass.
- A structure to skip setting and getting state if the spin control loop is running and trying to get to an orientation.
- A straightforward configuration for the control loop within the velocities file. This can be expanded.

Further work (not in this pr):
- Make the configuration of the loop easier within rdk by not relying on the attribute map construction - I'll look into this.
- Making sure we're setting and getting states as expected -  testing expansion on  `control_loop_test.go` and looking at the `makeSignal` function.
- Hardware tests on viam-base.

File: control/endpoint.go [hunk 8]
```diff
@@ -26,23 +26,29 @@ func newEndpoint(config BlockConfig, logger golog.Logger, ctr Controllable) (Blo
 }
 
 func (e *endpoint) Next(ctx context.Context, x []*Signal, dt time.Duration) ([]*Signal, bool) {
+	e.logger.Infof("z length %v", len(x))
+	e.logger.Infof("controllable is %v", e.ctr)
```
Comment (randhid) [inline control/endpoint.go:30]: These seem to be always nil

File: control/endpoint.go [hunk 9]
```diff
@@ -26,23 +26,29 @@ func newEndpoint(config BlockConfig, logger golog.Logger, ctr Controllable) (Blo
...
 	switch len(x) {
-	case 1:
-		power := x[0].GetSignalValueAt(0)
+	case 1, 2:
 		if e.ctr != nil {
-			err := e.ctr.SetState(ctx, power)
+			e.logger.Infof("setting state %v", x)
```
Comment (randhid) [inline control/endpoint.go:34]: keeping the logs as these will help in hardware testing.

File: components/base/sensorcontrolled/velocities.go [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+		// if the spin loop is polling, don't call set velocity, immediately return
+		// this allows us to keep the control loop unning without stopping it until
+		// the resource Close has been called
+		sb.logger.Info("skipping set state call")
+		return nil
+	}
+
+	sb.logger.Info("setting state")
+	linvel := state[0].GetSignalValueAt(0)
+	angvel := state[1].GetSignalValueAt(0)
```
Comment (randhid) [inline components/base/sensorcontrolled/velocities.go:122]: [q] Should I do this or add a helper for setting the state called in lines 55-61

File: components/base/sensorcontrolled/velocities.go [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+		sb.pollsensors(sensorCtx, extra)
+		return nil
+	}
+
+	sb.logger.Info("setting velocity without loop")
+	// else do not use the control loop and pass through the SetVelocity command
+	return sb.controlledBase.SetVelocity(ctx, linear, angular, extra)
+}
+
+func (sb *sensorBase) pollsensors(ctx context.Context, extra map[string]interface{}) {
```
Comment (randhid) [inline components/base/sensorcontrolled/velocities.go:68]: This function can go away now or later - it is a helper that polls the linear and angular velocity sensors from within the sensor-controlled base. I'm thinking of just having it run based on the use control loop and removing the check for `polling`. The `isPolling` functionality is related to stopping the `Spin` control loop (not using the control package) from activity when other API calls are made.

File: components/base/sensorcontrolled/sensorcontrolled.go (hunk 2 above)
Comment (randhid) [reply components/base/sensorcontrolled/sensorcontrolled.go]: There is now one star and one stop - we will deal with the loop activity in the SetState and State functions

File: components/base/sensorcontrolled/sensorcontrolled.go (hunk 1 above)
Comment (randhid) [reply components/base/sensorcontrolled/sensorcontrolled.go:179]: Kept this change in here.

File: components/base/sensorcontrolled/velocities.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (randhid) [reply components/base/sensorcontrolled/velocities.go]: Note: the diagram is no longer valid for the current loop - it is a single loop with velocity input and velocity outputs - no derivatives were used, a simple constant -> sum_block -> gain -> pid -> endpoint -> (back to) sum_block structure is now used. Drawing incoming.

File: control/control_loop.go [hunk 12]
```diff
@@ -195,7 +195,7 @@ func (l *Loop) Start() error {
 	if len(l.ts) == 0 {
 		return errors.New("cannot start the control loop if there are no blocks depending on an impulse")
 	}
-	l.logger.Debugf("Running loop on %1.4f %+v\r\n", l.cfg.Frequency, l.dt)
+	// l.logger.Infof("Running loop on %1.4f %+v\r\n", l.cfg.Frequency, l.dt)
```
Comment (martha-johnston) [inline control/control_loop.go:198]: nit: this could be helpful to have for debug

**Review by npmenard — APPROVED** [review]
LGTM

**Comment by randhid** [comment]
@martha-johnston added comments - let me know if there are more areas that could use some love.
@gvaradarajan take a look when you're ready!
